### PR TITLE
SegmentAnything #15-add sam replay strategy

### PR DESCRIPTION
**Pull request by jesicasusanto**
#15 
Summary
This pull request introduces a new replay strategy that utilizes the "Segment Anything with Masking" (SAM) algorithm to segment screenshots before performing optical character recognition (OCR) to generate text. The replay strategy then generates the next input event using the processed text and previously recorded input events.

Next step
Generate next InputEvent based on the predicted properties from the GPT-J model

Questions

What are the expected limitations of using the SAM algorithm for segmentation?
Are there any potential issues with using the GPT-J model for generating input events in this context?

I am happy to receive any feedback on this approach. Thanks!

Step by step instructions for checking out and testing :
```
git remote add jesicasusanto https://github.com/jesicasusanto/PAT.git
git fetch jesicasusanto
git checkout add-sam-replay-strategy
pip install -r requirements.txt 
python -m puterbot.replay SamReplayStrategy
```

### CODE REVIEW

**Comment by abrichr** [comment]
Thank you for getting the ball rolling on this @jesicasusanto !

> What are the expected limitations of using the SAM algorithm for segmentation?

The model itself is SOTA in segmentation. Manual testing in their web UI suggests it's great at segmenting UI elements (e.g. file menus, buttons, text inputs), as long as it has some user input to indicate the object of interest (which we can easily get from the user's click locations).

The bigger question in my view is how we can use the model to maximum effect. The first order usage is likely just using the model as a black box for segmenting the UI, as you have outlined here. However, as you build this out, it may be helpful to keep in mind that we may also want to share model parameters between vision and language models, for example as implemented in https://minigpt-4.github.io/ (see https://github.com/MLDSAI/puterbot/issues/12).

> Are there any potential issues with using the GPT-J model for generating input events in this context?

I'm sure there are many 😅 However take a look at the latest `DemoReplayStrategy` as a starting point (it seems this PR is based off an old version of the codebase): https://github.com/MLDSAI/puterbot/blob/main/puterbot/strategies/demo.py

Also, here's some code to assist in downloading the SAM model:

```
...
from pathlib import Path
...

CHECKPOINT_URL_BASE = "https://dl.fbaipublicfiles.com/segment_anything/"
CHECKPOINT_URL_BY_NAME = {
    "default": f"{CHECKPOINT_URL_BASE}sam_vit_h_4b8939.pth",
    "vit_l": f"{CHECKPOINT_URL_BASE}sam_vit_l_0b3195.pth",
    "vit_b": f"{CHECKPOINT_URL_BASE}sam_vit_b_01ec64.pth",
}
MODEL_NAME = "default"
CHECKPOINT_DIR_PATH = "./checkpoints"

class SamReplayStrategy(...):

    def __init__(
        ...
        model_name=MODEL_NAME,
        checkpoint_dir_path=CHECKPOINT_DIR_PATH,
        ....
    ):
        ...
        self.sam_model = self._initialize_model(model_name, checkpoint_dir_path)
        ...

    def _initialize_model(self, model_name, checkpoint_dir_path):
        checkpoint_url = CHECKPOINT_URL_BY_NAME[model_name]
        checkpoint_file_name = checkpoint_url.split("/")[-1]
        checkpoint_file_path = Path(checkpoint_dir_path, checkpoint_file_name)
        if not Path.exists(checkpoint_file_path):
            Path(checkpoint_dir_path).mkdir(parents=True, exist_ok=True)
            logger.info(f"downloading {checkpoint_url=} to {checkpoint_file_path=}")
            urllib.request.urlretrieve(checkpoint_url, checkpoint_file_path)
        return sam_model_registry[model_name](checkpoint=checkpoint_file_path)
```

**Comment by abrichr** [comment]
Hi @jesicasusanto , almost there! Please follow the instructions at https://github.com/MLDSAI/puterbot/pull/55#issuecomment-1530757878  but modified for your fork, e.g.:

```
git remote add jesicasusanto https://github.com/jesicasusanto/puterbot.git
git fetch jesicasusanto
# what goes here to run your code? please create a branch update your PR to merge from that branch, e.g.:
git checkout feat/<your-feature-name>  # please fill this in
pip install -r requirements.txt  # please fill this in
python -m ...  # please fill this in
```

**Comment by abrichr** [comment]
@jesicasusanto thank you for the great work! Please grab some time on my calendar at your earliest convenience on Weds/Thurs this week: https://www.getclockwise.com/c/richard-abrich/quick-meeting

**Comment by abrichr** [comment]
Thank you for the updated instructions! I get the following error after `python -m puterbot.replay SamReplayStrategy`:
```
(.venv) abrichr@MacBook-Pro-2 puterbot % python -m puterbot.replay SamReplayStrategy
2023-05-02 01:54:36.262 | INFO     | __main__:replay:19 - strategy_name='SamReplayStrategy'
zsh: illegal hardware instruction  python -m puterbot.replay SamReplayStrategy
```

I have seen this error before with torch on M2 silicon.


@jesicasusanto can you please clarify your platform by pasting the output of the following command:

```
(.venv) $ python -c 'import sys; import platform; print(sys.platform, platform.machine(), platform.processor()); import cpuinfo; print(cpuinfo.get_cpu_info())' 
```

Thank you!

**Comment by jesicasusanto** [comment]
Thank you for your response. Here is the output of the command you asked me to run:

`win32 AMD64 Intel64 Family 6 Model 142 Stepping 12, GenuineIntel
{'python_version': '3.10.10.final.0 (64 bit)', 'cpuinfo_version': [9, 0, 0], 'cpuinfo_version_string': '9.0.0', 'arch': 'X86_64', 'bits': 64, 'count': 8, 'arch_string_raw': 'AMD64', 'vendor_id_raw': 'GenuineIntel', 'brand_raw': 'Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz', 'hz_advertised_friendly': '1.8000 GHz', 'hz_actual_friendly': '1.7920 GHz', 'hz_advertised': [1800000000, 0], 'hz_actual': [1792000000, 0], 'l2_cache_size': 1048576, 'stepping': 12, 'model': 142, 'family': 6, 'l3_cache_size': 8388608, 'flags': ['3dnow', '3dnowprefetch', 'abm', 'acpi', 'adx', 'aes', 'apic', 'avx', 'avx2', 'bmi1', 'bmi2', 'clflush', 'clflushopt', 'cmov', 'cx16', 'cx8', 'de', 'ds_cpl', 
'dtes64', 'dts', 'erms', 'est', 'f16c', 'fma', 'fpu', 'fxsr', 'ht', 'ia64', 'intel_pt', 'invpcid', 'lahf_lm', 'mca', 'mce', 'mmx', 'mon
itor', 'movbe', 'mpx', 'msr', 'mtrr', 'osxsave', 'pae', 'pat', 'pbe', 'pcid', 'pclmulqdq', 'pdcm', 'pge', 'pni', 'popcnt', 'pse', 'pse3
6', 'rdrnd', 'rdseed', 'sep', 'serial', 'sgx', 'smap', 'smep', 'ss', 'sse', 'sse2', 'sse4_1', 'sse4_2', 'ssse3', 'tm', 'tm2', 'tsc', 'tscdeadline', 'vme', 'vmx', 'x2apic', 'xsave', 'xtpr'], 'l2_cache_line_size': 256, 'l2_cache_associativity': 6}`

**Comment by abrichr** [comment]
Got it running by removing PaddleOCR. Here the error I'm encountering:

```
2023-05-02 16:13:26.058 | INFO     | __main__:replay:37 - strategy=<puterbot.strategies.sam.SamReplayStrategy object at 0x16f958c70>                                                                                                                      
Traceback (most recent call last):                                                                                           
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main                                                                                                        
    return _run_code(code, main_globals, None,                                                                               
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code                                                                                                                   
    exec(code, run_globals)                                                                                                  
  File "/Users/abrichr/MLDSAI/src/puterbot/puterbot/replay.py", line 43, in <module>                                                                                                                                                                      
    fire.Fire(replay)                                                                                                                                                                                                                                     
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/fire/core.py", line 141, in Fire                                                                                                                                            
    component_trace = _Fire(component, args, parsed_flag_args, context, name)                                                                                                                                                                             
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/fire/core.py", line 466, in _Fire                                                                                                                                           
    component, remaining_args = _CallAndUpdateTrace(                                                                                                                                                                                                      
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace                                                                                                                             
    component = fn(*varargs, **kwargs)                                                                                                                                                                                                                    
  File "/Users/abrichr/MLDSAI/src/puterbot/puterbot/replay.py", line 39, in replay                                                                                                                                                                        
    strategy.run()                                                                                                                                                                                                                                        
  File "/Users/abrichr/MLDSAI/src/puterbot/puterbot/strategies/base.py", line 47, in run                                                                                                                                                                  
    input_event = self.get_next_input_event(screenshot)                                                                                                                                                                                                   
  File "/Users/abrichr/MLDSAI/src/puterbot/puterbot/strategies/sam.py", line 81, in get_next_input_event                                                                                                                                                  
    self.sam_predictor.set_image(screenshot.array)                                                                           
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/segment_anything/predictor.py", line 154, in predict                                                                                                                        
    masks, iou_predictions, low_res_masks = self.predict_torch(                                                              
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context                                                                                                                  
    return func(*args, **kwargs)                                                                                                                                                                                                                          
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/segment_anything/predictor.py", line 222, in predict_torch                                                                                                                  
    sparse_embeddings, dense_embeddings = self.model.prompt_encoder(                                                                                                                                                                                      
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl                                                                                                                       
    return forward_call(*args, **kwargs)                                                                                                                                                                                                                  
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/segment_anything/modeling/prompt_encoder.py", line 162, in forward                                                                                                          
    dense_embeddings = self._embed_masks(masks)                                                                              
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/segment_anything/modeling/prompt_encoder.py", line 104, in _embed_masks                                                                                                     
    mask_embedding = self.mask_downscaling(masks)                                                                            
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl                                                                                                                       
    return forward_call(*args, **kwargs)                                                                                     
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/container.py", line 217, in forward                                                                                                                        
    input = module(input)                                                                                                                                                                                                                                 
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl                                                                                                                       
    return forward_call(*args, **kwargs)                                                                                                                                                                                                                  
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 463, in forward                                                                                                                             
    return self._conv_forward(input, self.weight, self.bias)                                                                                                                                                                                              
  File "/Users/abrichr/MLDSAI/src/puterbot/.venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 459, in _conv_forward                                                                                                                       
    return F.conv2d(input, weight, bias, self.stride,                                                                                                                                                                                                     
RuntimeError: Given groups=1, weight of size [4, 1, 2, 2], expected input[1, 1964, 3024, 3] to have 1 channels, but got 1964 channels instead                                
```

@jesicasusanto have you seen anything like this before? Any ideas?

**Comment by jesicasusanto** [comment]
Thank you for always giving feedback as it helps me to reach my goal for this issue. I believe I have never seen that error before. 
However, I believe it is caused by 
`masks, score, logit = self.sam_predictor.predict(mask_input=screenshot.array, multimask_output=False)
There is a mismatch between the expected number of channels in the input image and the number of channels provided.

**Comment by jesicasusanto** [comment]
Regarding the SamReplayStrategy class I'm currently working on, I have a question regarding the generation of the InputEvent object. I would appreciate your guidance on whether the GPT model is expected to generate all the attributes of the InputEvent especially the parent ID, child InputEvents, recording, screenshot, and window event

I apologize for any confusion, but I am uncertain about the capabilities of the model in this regard. Your clarification would be greatly appreciated. Thank you!

**Comment by jesicasusanto** [comment]
I've decided to work on [https://github.com/MLDSAI/OpenAdapt/issues/168](url) instead.

File: openadapt/strategies/sam.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        #Segment the Screenshot with SAM
+        image = screenshot.array
+        masks = self.mask_generator.generate(image)
+        plt.figure(figsize=(20,20))
+        plt.imshow(image)
+        self.show_anns(masks)
+        plt.axis('off')
+        plt.show()
+
+
```
Comment (abrichr) [inline openadapt/strategies/sam.py:95]: Please format with python black to remove unnecessary newlines

**Pull request closed without merging**